core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -0,0 +1,211 @@
+  }
+
+  private[server] def sendRequest(request: RequestChannel.Request,
+                                  callback: RequestCompletionHandler): Unit = {
+    val requestBuilder = new NoOpRequestBuilder(request.context.header.apiKey, request.body[AbstractRequest])
+    requestQueue.put(BrokerToControllerQueueItem(requestBuilder, callback, ""))
+  }
+
+  private[server] def forwardRequest(originalRequest: RequestChannel.Request,
+                                     callback: RequestCompletionHandler): Unit = {
+    val requestBuilder = new NoOpRequestBuilder(originalRequest.context.header.apiKey, originalRequest.body[AbstractRequest])
+    requestQueue.put(BrokerToControllerQueueItem(requestBuilder, callback, originalRequest.context.principal.getName))
+  }
+}
+
+case class BrokerToControllerQueueItem(request: AbstractRequest.Builder[_ <: AbstractRequest],
+                                       callback: RequestCompletionHandler,
+                                       initialPrincipalName: String)
+
+class BrokerToControllerRequestThread(networkClient: KafkaClient,
+                                      metadataUpdater: ManualMetadataUpdater,
+                                      requestQueue: BlockingQueue[BrokerToControllerQueueItem],
+                                      metadataCache: kafka.server.MetadataCache,
+                                      config: KafkaConfig,
+                                      listenerName: ListenerName,
+                                      time: Time,
+                                      threadName: String)
+  extends InterBrokerSendThread(threadName, networkClient, time, isInterruptible = false) {
+
+  private var activeController: Option[Node] = None
+
+  override def requestTimeoutMs: Int = config.controllerSocketTimeoutMs
+
+  override def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+    val requestsToSend = new mutable.Queue[RequestAndCompletionHandler]
+    if (requestQueue.peek() != null) {
+      val topRequest = requestQueue.take()
+
+      val request = RequestAndCompletionHandler(
+        activeController.get,
+        topRequest.request,
+        handleResponse(topRequest),
+        topRequest.initialPrincipalName)
+      requestsToSend.enqueue(request)
+    }
+    requestsToSend
+  }
+
+  private[server] def handleResponse(request: BrokerToControllerQueueItem)(response: ClientResponse): Unit = {
+    if (response.wasDisconnected()) {
+      activeController = None
+      requestQueue.put(request)
+    } else if (response.responseBody().errorCounts().containsKey(Errors.NOT_CONTROLLER)) {
+      // just close the controller connection and wait for metadata cache update in doWork
+      networkClient.close(activeController.get.idString)
+      activeController = None
+      requestQueue.put(request)
+    } else {
+      request.callback.onComplete(response)
+    }
+  }
+
+  private[server] def backoff(): Unit = pause(100, TimeUnit.MILLISECONDS)
```

Comment:
Maybe an exponential backoff instead of constant?

Reply:
Will consider adding it after https://github.com/apache/kafka/pull/8421 lands